Added a DelegateMvRxView

tunjid opened the pull request:
A simple Delegate that uses the saved state artifact, for cases where inheriting from `BaseMvRxFragment` is unfeasible, undesirable, or for composing with a custom view. The custom view may choose to grab the `SavedStateRegistryOwner` from either an `Activity` or `Fragment`.

I was able to test this with DogsFragment fine, even across process death. I'm not sure how to contribute this change with tests for this project however, pointers are much appreciated.

There's also the function delegates that will need to be added for `activityViewModel` and the like should this initial change be worthwhile.

Usage was as follows:

```kotlin
class DogsFragment : Fragment(), DogsFragmentHandler {

    private val view = DelegateMvRxView(this, ::invalidate)
    private val viewModel: DogViewModel by activityViewModel(view)
    private lateinit var bindings: DogsFragmentBinding
    private val adapter = DogAdapter(this)

    ...

    private fun invalidate() = withState(viewModel) { state ->
        bindings.state = state
    }
}
```

tunjid commented on mvrx/src/main/kotlin/com/airbnb/mvrx/DelegateMvRxView.kt:
```diff
@@ -0,0 +1,39 @@
+import androidx.savedstate.SavedStateRegistryOwner
+
+class DelegateMvRxView(
+        private val owner: SavedStateRegistryOwner,
```
This can be either an `Activity` or `Fragment` Should Google. The type could be more open to allow. for a different method of `mvrxViewId` saving and restoration.

elihart replied:
this looks pretty reasonable to me. can you add some kdoc to this class to describe the purpose and when it may be used?

tunjid commented:
I added a `by activityViewModel(view)` when I was testing, but I removed it from the PR as it was a copy paste of the existing `activityViewModel` extension on a `Fragment`; I wanted to make sure the PR had legs before adding more methods and generally present a small diff at first.

I'll go ahead and use the Delegate in the `BaseMvRxFragment`.

elihart commented on mvrx/src/main/kotlin/com/airbnb/mvrx/BaseMvRxFragment.kt:
```diff
@@ -12,30 +12,25 @@ import androidx.lifecycle.LifecycleOwner
-        // subscribe to a ViewModel.
-        postInvalidate()
-    }
+    override fun invalidate() = Unit
```
shouldn't we leave this unimplemented to force base classes to implement?

elihart commented on mvrx/src/main/kotlin/com/airbnb/mvrx/BaseMvRxFragment.kt:
```diff
@@ -12,30 +12,25 @@ import androidx.lifecycle.LifecycleOwner
     override fun onCreate(savedInstanceState: Bundle?) {
-        mvrxViewIdProperty.restoreFrom(savedInstanceState)
         super.onCreate(savedInstanceState)
+        viewDelegate = DelegateMvRxView(
```
might be nice to provide a fragment specific implementation of this to make it easier to drop in for the common case

tunjid replied:
Sounds good, will add an extension for it.

elihart commented on mvrx/src/main/kotlin/com/airbnb/mvrx/BaseMvRxFragment.kt:
```diff
@@ -12,30 +12,22 @@ import androidx.lifecycle.LifecycleOwner
-        mvrxViewIdProperty.restoreFrom(savedInstanceState)
         super.onCreate(savedInstanceState)
+        viewDelegate = delegateMvRxView(
+                subscriptionLifecycleOwner = subscriptionLifecycleOwner,
```
I think `subscriptionLifecycleOwner` is not necessary to explicitly define now, as the default is the same?

elihart commented on mvrx/src/main/kotlin/com/airbnb/mvrx/DelegateMvRxView.kt:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Convenience for providing a [DelegateMvRxView] to a [Fragment]
+ */
+fun Fragment.delegateMvRxView(
```
nice :)

gpeal commented:
@tunjid This should no longer be necessary as of https://github.com/airbnb/MvRx/pull/396 :)

Thanks for suggesting we take another look at this! It requried some new APIs that didn't exist when MvRx was first built (ViewModel SavedStateHandle).
